api/pkg/apis/v1alpha1/managers/campaigncontainers/campaign-container-manager.go

```diff
@@ -0,0 +1,181 @@
+
+	err = t.StateProvider.Delete(ctx, states.DeleteRequest{
+		ID: name,
+		Metadata: map[string]interface{}{
+			"namespace": namespace,
+			"group":     model.WorkflowGroup,
+			"version":   "v1",
+			"resource":  "campaigncontainers",
+			"kind":      "CampaignContainer",
+		},
+	})
+	return err
+}
+
+func (t *CampaignContainersManager) UpsertState(ctx context.Context, name string, state model.CampaignContainerState) error {
+	ctx, span := observability.StartSpan("CampaignContainersManager", ctx, &map[string]string{
+		"method": "UpsertState",
+	})
+	var err error = nil
+	defer observ_utils.CloseSpanWithError(span, &err)
+
+	if state.ObjectMeta.Name != "" && state.ObjectMeta.Name != name {
+		return v1alpha2.NewCOAError(nil, fmt.Sprintf("Name in metadata (%s) does not match name in request (%s)", state.ObjectMeta.Name, name), v1alpha2.BadRequest)
+	}
+	state.ObjectMeta.FixNames(name)
+
+	body := map[string]interface{}{
+		"apiVersion": model.WorkflowGroup + "/v1",
+		"kind":       "CampaignContainer",
+		"metadata":   state.ObjectMeta,
+		"spec":       state.Spec,
+	}
+
+	upsertRequest := states.UpsertRequest{
+		Value: states.StateEntry{
+			ID:   name,
+			Body: body,
+			ETag: "",
+		},
+		Metadata: map[string]interface{}{
+			"namespace": state.ObjectMeta.Namespace,
+			"group":     model.WorkflowGroup,
+			"version":   "v1",
+			"resource":  "campaigncontainers",
+			"kind":      "CampaignContainer",
+		},
+	}
+	_, err = t.StateProvider.Upsert(ctx, upsertRequest)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (t *CampaignContainersManager) ListState(ctx context.Context, namespace string) ([]model.CampaignContainerState, error) {
+	ctx, span := observability.StartSpan("CampaignContainersManager", ctx, &map[string]string{
+		"method": "ListState",
+	})
+	var err error = nil
+	defer observ_utils.CloseSpanWithError(span, &err)
+
+	listRequest := states.ListRequest{
+		Metadata: map[string]interface{}{
+			"version":   "v1",
+			"group":     model.WorkflowGroup,
+			"resource":  "campaigncontainers",
+			"namespace": namespace,
+			"kind":      "CampaignContainer",
+		},
+	}
+	var campaigncontainers []states.StateEntry
+	campaigncontainers, _, err = t.StateProvider.List(ctx, listRequest)
+	if err != nil {
+		return nil, err
+	}
+	ret := make([]model.CampaignContainerState, 0)
+	for _, t := range campaigncontainers {
+		var rt model.CampaignContainerState
+		rt, err = getCampaignContainerState(t.Body, t.ETag)
+		if err != nil {
+			return nil, err
+		}
+		ret = append(ret, rt)
+	}
+	return ret, nil
+}
+
+func getCampaignContainerState(body interface{}, etag string) (model.CampaignContainerState, error) {
+	var CampaignContainerState model.CampaignContainerState
+	bytes, _ := json.Marshal(body)
+	err := json.Unmarshal(bytes, &CampaignContainerState)
+	if err != nil {
+		return model.CampaignContainerState{}, err
+	}
+	if CampaignContainerState.Spec == nil {
+		CampaignContainerState.Spec = &model.CampaignContainerSpec{}
+	}
```

Comment:
I would recommend initializing nil Status field as well for all the containerState